### ISSUE TITLE
Mechanism to track provenance of output records

### ISSUE THREAD

**Issue by ryzhyk** [ryzhyk/differential-datalog#103]

**Comment by ryzhyk** [comment]
One way to do this is to add a provenance field to `Value`, but make it invisible to comparison operators, so that different derivations of the same Value look the same to differential. This way provenance won't change the semantics of DDlog, but we will only observe one derivation of each value.

**Comment by namibj** [comment]
I just want to mention the existence of prior art: https://github.com/frankmcsherry/explanation
It is only for an ancient version of Differential Dataflow, but lessons learnt should be valuable

**Comment by sdifv** [comment]
> One way to do this is to add a provenance field to `Value`, but make it invisible to comparison operators, so that different derivations of the same Value look the same to differential. This way provenance won't change the semantics of DDlog, but we will only observe one derivation of each value.

Indeed, this is a very intuitive method, but as you said, making this field invisible will actually make the provenance information partly missing if there existing multi-derivation rule.

I found that another datalog dialect, **souffle’**, already supports the [provenance feature](https://souffle-lang.github.io/provenance#the-explain-interface) to help users with interactive debugging. Could you please share me whether there is any development plan for DDlog in this aspect? Therefore, users do not need to directly modify the DDlog code logic to get the support of provenance.

Thank you!

**Comment by mbudiu-vmw** [comment]
This is certainly something we'd like to do, but at this point no one is working on this topic.

**Comment by Kixiron** [comment]
Soufflé has done some other [relevant work](https://youtu.be/4aXy9UGq4Ng) on this recently
